metadatastore/test/test_mds.py

```diff
@@ -1,11 +1,401 @@
+import time as ttime
+import uuid
+import pytest
+import warnings
+import datetime
 
 from metadatastore.mds import MDS
+from metadatastore.core import (NoRunStart, NoEventDescriptors)
 
 
-def test_pickle():
+def check_for_id(document):
+    """Make sure that our documents do not have an id field
+
+    Parameters
+    ----------
+    document : metadatastore.document.Document
+        A sanitized mongoengine document
+    """
+    with pytest.raises(KeyError):
+        document['id']
+
+
+def setup_syn(mds, custom=None):
+    if custom is None:
+        custom = {}
+    data_keys = {k: {'source': k,
+                     'dtype': 'number',
+                     'shape': None} for k in 'ABCEDEFGHIJKL'
+                 }
+    scan_id = 1
+
+    # Create a BeginRunEvent that serves as entry point for a run
+    rs = mds.insert_run_start(scan_id=scan_id, beamline_id='testing',
+                              time=ttime.time(),
+                              uid=str(uuid.uuid4()),
+                              **custom)
+
+    # Create an EventDescriptor that indicates the data
+    # keys and serves as header for set of Event(s)
+    e_desc = mds.insert_descriptor(data_keys=data_keys,
+                                   time=ttime.time(),
+                                   run_start=rs, uid=str(uuid.uuid4()))
+    return rs, e_desc, data_keys
+
+
+def syn_data(data_keys, count):
+
+    all_data = deque()
+    for seq_num in range(count):
+        data = {k: float(seq_num) for k in data_keys}
+        timestamps = {k: ttime.time() for k in data_keys}
+
+        _time = ttime.time()
+        uid = str(uuid.uuid4())
+        all_data.append({'data': data, 'timestamps': timestamps,
+                         'seq_num': seq_num, 'time': _time,
+                         'uid': uid})
+    return all_data
+
+
+@pytest.mark.parametrize('ver', (0, 1))
+def test_pickle(ver):
     md = MDS(config={'host': 'portland'}, version=1)
     md2 = pickle.loads(pickle.dumps(md))
 
     assert md.version == md2.version
     assert md.config == md2.config
+
+
+def test_event_descriptor_insertion(mds_all):
+    mds = mds_all
+    # format some data keys for insertion
+    data_keys = {'some_value': {'source': 'PV:pv1',
+                                'shape': [1, 2],
+                                'dtype': 'array'},
+                 'some_other_val': {'source': 'PV:pv2',
+                                    'shape': [],
+                                    'dtype': 'number'},
+                 'data_key3': {'source': 'PV:pv1',
+                               'shape': [],
+                               'dtype': 'number',
+                               'external': 'FS:foobar'}}
+
+    document_insertion_time = ttime.time()
+    run_start_uid = mds.insert_run_start(scan_id=3022013,
+                                         beamline_id='testbed',
+                                         owner='tester',
+                                         group='awesome-devs',
```

Comment:
On Sat, 7 May 2016, Dan Allan wrote:
...
> On Sat, 7 May 2016, Li Li wrote:
> > +    run_start_uid = mds.insert_run_start(scan_id=3022013,
> > +                                         beamline_id='testbed',
> > +                                         owner='tester',
> > +                                         group='awesome-devs',
...
> > When reading some databroker information, I find owner/group
> > information is probably not used correctly. The owner is the
> > beamline scientist, and group is the user group? Sorry for the
> > tedious questions.
> 
> There is no established standard. Ultimately, this kind of info
> should match up with the information in a beam time proposal. There
> will be a standard some day. But for now it's up to users and
> beamline scientists to enter useful, accurate info. We have no way
> of verifying it at this time.


To make things more clear, we could change 'group' to 'proposal' and
'owner' to 'collector'.

A proposal is the standard unit of access control at nsls2 and most
facilities, so we need the proposal.  The pi of the proposal
ultimately owns the data, so with the proposal, the actual 'owner'
(the pi) would be redundant.

The pi is very often not the person collecting data though and it
makes sense to record who does.


We might not have a documented standard or connection to pass (yet),
but the basic needs and requirements are clear, not remotely new, and
there are examples at a bunch of facilities.

Until we are collecting data using individual accounts (either at the
os or application level) and have a link to pass, it will still be
up to users and beamline scientists to enter useful, accurate info,
but we do know what fields they should be entering.

-matt